api/src/main/java/jakarta/data/repository/CrudRepository.java

```diff
@@ -0,0 +1,103 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package jakarta.data.repository;
+
+import jakarta.data.exceptions.EntityExistsException;
+
+/**
+ * <p>A repository interface that extends the capabilities of basic operations on entities, including insert and update operations.</p>
+ *
+ * <p>This repository extends the {@link BasicRepository} interface, providing a comprehensive set of methods to interact with
+ * persistent entities of type {@code <T>}, where {@code <T>} represents the entity bean type, and {@code <K>} represents the key type.</p>
+ *
+ * <p>It encompasses standard CRUD (Create, Read, Update, Delete) operations, allowing you to perform insert and update operations in
+ * addition to basic retrieval and deletion. This interface combines the Data Access Object (DAO) aspect with the repository pattern,
+ * offering a versatile and complete solution for managing persistent entities within your Java applications.</p>
+ *
+ * @param <T> the entity bean type
+ * @param <K> the key type.
+ * @see BasicRepository
+ * @see DataRepository
+ */
+public interface CrudRepository<T, K> extends BasicRepository<T, K> {
+
+    /**
+     * <p>Inserts an entity into the database. If an entity of this type with the same
+     * unique identifier already exists in the database, then this method raises
+     * {@link EntityExistsException} for.</p>
+     *
+     * @param entity the entity to insert. Must not be {@code null}.
+     * @return the inserted entity.
+     * @throws EntityExistsException if the entity is already present in the database.
+     * @throws NullPointerException if the entity is null.
+     * @throws UnsupportedOperationException for databases that use an append model to write data.
+     */
+    T insert(T entity);
+
+    /**
+     * <p>Inserts multiple entities into the database. If an entity of this type with the same
+     * unique identifier as any of the given entities already exists in the database,
+     * then this method raises {@link EntityExistsException}.</p>
+     *
+     * @param entities entities to insert.
+     * @return an iterable containing the inserted entities.
+     * @throws EntityExistsException if any of the entities are already present in the database.
+     * @throws NullPointerException if either the iterable is null or any element is null.
+     * @throws UnsupportedOperationException for Key-Value and Wide-Column databases
+     *         that use an append model to write data.
+     */
+    Iterable<T> insertAll(Iterable<T> entities);
```

Comment:
The return value should not be switched to the inserted entities, which would require additional retrieval from the database. When users want the pattern of performing the additional retrieval, they will use saveAll, not insertAll.

```suggestion
     * @throws EntityExistsException if any of the entities are already present in the database.
     * @throws NullPointerException if either the iterable is null or any element is null.
     * @throws UnsupportedOperationException for Key-Value and Wide-Column databases
     *         that use an append model to write data.
     */
    void insertAll(Iterable<T> entities);
```

Reply:
Why? I mean, with SQL I can return the same values from the parameter, it might be fine, isn't it?

The point is with Record:

So, how do we handle records in any persistence capability?

```java
record Book(@AutoIncrement @Id Long id, String name){
}
var book = Book.of(null, "Effective Java");
library.insert(book);

````

Reply:
> Why? I mean, with SQL I can return the same values from the parameter, it might be fine, isn't it?
> 
> The point is with Record:
> 
> So, how do we handle records in any persistence capability?
> 
> ```java
> record Book(@AutoIncrement @Id Long id, String name){
> }
> var book = Book.of(null, "Effective Java");
> library.insert(book);
> ```

I'm pretty sure you asked this same question elsewhere and I answered it, but I'll do so here as well.  Record entities vs Class entities and SQL vs NoSQL makes absolutely no difference here and will have the same problems either way.  What matters for this scenario is the fact that you requested an automatically generated id.  You don't know what id was generated unless the method returns it to you.  In Jakarta Data, we have a method that does exactly that: `save`, and that's what you should be using for this pattern.  Querying the database is not what `insert` is for.

Reply:
I will close this one here once we agree to go deep on this in the next PR.